### DISCUSSION TITLE
Share a state

### DISCUSSION THREAD

**Discussion by josejachuf** [salvo-rs/salvo#69] [Q&A]
Hi, I'm starting to learn Rust, particularly watching some web frameworks. I liked Tide, but yesterday I knew Salvo and wanted to migrate some things in which I was working to Salvo.

In Tide I can share a state enter all the requests, a pool of connections for example 

`let mut app = tide::with_state(state);`

How do I inject my pool of connections with Salvo?
Use a Middleware? If so, how do I do it? 

I use sqlx

```
use sqlx::{Pool, PgPool};


pub struct State {
    db_pool: PgPool,
}

pub async fn make_db_pool(db_url: &str) -> PgPool {
    Pool::connect(&db_url).await.unwrap()
}
```

Thanks
Jose

**Comment by fergus-hou** [comment]
You can use some code like this:
```
use sqlx::{Pool, PgPool};
use once_cell::sync::OnceCell;

pub static DB_POOL: OnceCell<PgPool> = OnceCell::new();

pub fn db_pool() -> &PgPool {
    DB_POOL.get().unwrap()
}

pub async fn make_db_pool(db_url: &str) -> PgPool {
    Pool::connect(&db_url).await.unwrap()
}

#[tokio::main]
async fn main() {
    let pool = make_db_pool().await;
    DB_POOL.set(pool).unwrap();
}
```
Just call db_pool when you want pool. 

You can reference this if you use diesel:
https://github.com/TechEmpower/FrameworkBenchmarks/blob/master/frameworks/Rust/salvo/src/main_db.rs
https://github.com/TechEmpower/FrameworkBenchmarks/blob/master/frameworks/Rust/salvo/src/main_pg.rs

**Comment by josejachuf** [comment]
I will see this. Thanks @fergus-hou
